### ISSUE TITLE
Add examples for skorch

### ISSUE THREAD

**Issue by UTUnex** [facebook/Ax#367]
Hi, I'm using skorch, which is a wrapper of pytorch to make it have a sklearn-API for easy use, for NN training and I hope to use ax for hyperparameters tuning in skorch. After checking your turorial, I still feel very hard to start, could you please add an example for ax using with skorch? 
Or at least, please add an example for sklearn with ax.

Thank you.

**Comment by lena-kashtelyan** [comment]
Hi, @UTUnex, we'll definitely note that feature request for future work! In the meantime, maybe you could provide us with a bit more information on what parameters you are trying to optimize and what the search space looks like for you? 

Also, which tutorials did you check out? I wonder if [our `pytorch` tutorial](https://ax.dev/tutorials/tune_cnn.html) could be of help? That tutorial uses our Loop API (`optimize`), but you could also opt for our `Service API` instead ([tutorial](https://ax.dev/tutorials/gpei_hartmann_service.html)), which just provides you with parameter configurations to try and allows you to evaluate them in any fitting way, then logging the data back to Ax. That API is very independent of how you need to evaluate the trials (parameter configurations) as part of your experiment, so it might be easier for your very start with Ax.

**Comment by UTUnex** [comment]
Hi,
Thank your for your reply. Please allow me to talk a bit more about what I expect and what I have tried (Please be patient to read this):


(1)   I use Jupyter notebook for all my ML and DL task ,so all things here I talk about are based on the code in the Jupyter notebook not in the *.py format as they have different style and logic.

(2)   In my mind, the DL code should have at least four sections : Data loading and preprocessing section ; Neural net design section; neural net training and tuning section; and the result logging + visualization + monitoring section

(3)   Among the four sections, I only hope to put more energy on the neural net design section which means what kind of layers I should import and how many of them and how many nodes in each of them I should offer in my net; as for the rest parts, I always expect for some high-level wrappers which provide many parameters and clearly describe what they mean, and so what I need to do is just provide the required parameters and let it go.

(4)   The Jupyter notebook has helped me get out of many ‘def’s as it offers the cell to wrap and segment codes so I don’t have to write that many python functions (ofc sometimes they are still necessary); the scorch has greatly helped me with the training part as with it I don’t have to write those ‘for loops’ or more complicated training functions, I can use it just like a sklearn classfier/regressor and I just need consider what parameters I should provide to this classfier/regressor.


(5)   Now, the bottleneck comes. I have spent many days looking for a wrapper tuner and apply it into my scorch network but it’s a pity that until now, none of the tools I have found and tried can meet my requirement. I don’t mind talking a bit more about them:



n  Ray tune: it has rich searching algorithms and early stopping schedulars but it’s not that easy to start because you need consider many things outside of the neural net itself and you have to set them one by one.  It indeed has a good wrapper for sklearn or sklearn-like libraries called tune-sklearn but its documentation support is very bad and it seems they are not putting much energy on this branch so you know the ‘tune-sklearn’ is not an active project. But I just wanna say, the tuner format that tune-sklearn offers is just the format that I expect for. It seems, their tune-sklearn is just a semi-finished product…

Well, if possible, I really hope in the short future ax can provide a AXsearchCV() or something like that which can be easily merged in my code.

n  Microsoft nni: it seems it doesn’t support code in jupyter notebook and only supports *,py script. It indeed provides good examples for sklearn which I can refer to for my skorch work (yeah they are still different but at least I can start). The problem is when I tried their examples on my own PC, many errors occurred and even with some adjustment after checking their GitHub issues, my first example still cannot run, yeah those errors are still there…So you see, I can’t even start,,,,

n  Wandb-sweep: honestly, I indeed put much hope on this product when I found it offers Bayes-optimization plus hyperband early stopping scheduler plus DL process visualization and monitoring functions.. yeah it looks promising, but just like nni, I failed my first example and in their documentation, I cannot find any detailed examples using skorch, and currently their functions have bugs when using jupyter notebook, so yeah, it cannot solve my problem..

n  Optuna, I also submit a feature adding issue to them and they just started the work, so yeah currently it’s not applicable.



(6)   So what I expect is :

n  Support Jupyter notebook which allows me to use the tuner in a cell

n  Has a sklearn gridsearchCV or randomsearchCV format but contains more advanced algorithms and early stopping schedulars, which means, wrap those advanced algorithms and schedulars under the hood and just provide the user a skleran-like API, for example, Ax_searchCV(optimizer = ‘bohb’, early_stopping_scheduler = ‘bohb’) Or Ax_searchCV(optimizer = ‘TPE’, early_stopping_scheduler = ‘hyperband’), yeah something like this.

n  Support single and multiple GPU usage, for example, Ax_searchCV(GPU = True, GPU_num = 4)

n  Support visualization and process monitoring, for example, Ax_searchCV(logger = ‘Ax_board’) or Ax_searchCV(logger = ‘wandb’) or Ax_searchCV(logger = ‘tensorboard’)..

n  Yeah, what I like is, you put all the required parameters in the wrapper Ax_searchCV( ) and we just need set any of them in the wrapper instead of consider how to build them ourselves, so it’s a highly-wrapped and quite functional tuner.

n  As for the hyperparameters format, yml or python dict are both OK for me.

Just for your reference but I really hope you can make it very soon.

Thank you.


发送自 Windows 10 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>应用

发件人: Lena Kashtelyan<mailto:notifications@github.com>
发送时间: 2020年8月25日 16:42
收件人: facebook/Ax<mailto:Ax@noreply.github.com>
抄送: Unex<mailto:hsingliangchao@hotmail.com>; Mention<mailto:mention@noreply.github.com>
主题: Re: [facebook/Ax] Add examples for skorch (#367)


Hi, @UTUnex<https://github.com/UTUnex>, we'll definitely note that feature request for future work! In the meantime, maybe you could provide us with a bit more information on what parameters you are trying to optimize and what the search space looks like for you?

Also, which tutorials did you check out? I wonder if our pytorch tutorial<https://ax.dev/tutorials/tune_cnn.html> could be of help? That tutorial uses our Loop API (optimize), but you could also opt for our Service API instead (tutorial<https://ax.dev/tutorials/gpei_hartmann_service.html>), which just provides you with parameter configurations to try and allows you to evaluate them in any fitting way, then logging the data back to Ax. That API is very independent of how you need to evaluate the trials (parameter configurations) as part of your experiment, so it might be easier for your very start with Ax.

―

**Comment by lena-kashtelyan** [comment]
I don't think examples for skorch are on our radar in the short-term, but we will reopen this issue if they are prioritized!
